src/editor/codemirror/language-server/diagnostics.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * (c) 2021, Micro:bit Educational Foundation and contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+import { Diagnostic } from "@codemirror/lint";
+import { Text } from "@codemirror/text";
+import * as LSP from "vscode-languageserver-protocol";
+import { positionToOffset } from "./positions";
+
+const severityMapping = {
+  [LSP.DiagnosticSeverity.Error]: "error",
+  [LSP.DiagnosticSeverity.Warning]: "warning",
+  [LSP.DiagnosticSeverity.Information]: "info",
+  [LSP.DiagnosticSeverity.Hint]: "info",
+} as const;
+
+export const diagnosticsMapping = (
+  document: Text,
+  diagnostics: LSP.Diagnostic[]
+): Diagnostic[] =>
+  diagnostics
+    .map(({ range, message, severity }): Diagnostic | undefined => {
+      let from = positionToOffset(document, range.start);
+      let to = positionToOffset(document, range.end);
+      console.log(message, range.start, range.end, from, to);
+
+      // Skip if we can't map to the current document.
+      if (from !== undefined && to !== undefined) {
+        // If `from` is the end of one line and `to` is the beginning of another
+        // then CM won't display the diagnostic inline. Shift the `to` to also
+        // be at the end of the line.
```

Comment:
Add a note that we can remove once the commit referenced here is merged: https://discuss.codemirror.net/t/diagnostics-for-the-range-from-end-of-a-line-to-the-start-of-the-next/3495